languageserver/server/server.go

```diff
@@ -1953,8 +1959,23 @@ func (s *Server) getDiagnostics(
 		diagnostics = append(diagnostics, extraDiagnostics...)
 	}
 
-	for _, hint := range checker.Hints() {
-		diagnostic, codeActionsResolver := convertHint(hint, uri)
+	if checkError != nil {
+		return
+	}
+
+	analysisProgram := analysis.Program{
+		Program:     program,
+		Elaboration: checker.Elaboration,
+		Location:    checker.Location,
+		Code:        text,
+	}
+
+	var reportLock sync.Mutex
+
+	report := func(linterDiagnostic analysis.Diagnostic) {
+		reportLock.Lock()
+		defer reportLock.Unlock()
```

Comment:
Is this lock required because the analysis below runs concurrently?

Reply:
Yeah, it looks like it based on the implementation in `tools/analysis/program.go` as well as this being how the linter itself implements its report function.

Reply:
Probably not a big deal, but there might be some contention here. We might want to change `analysis.Program.Run` to not accept a single reporting function that is used for all diagnostics of all given analyzers, but report diagnostics for each analyzer separately, like Go does: https://github.com/golang/tools/blob/d6c099e3c1a39214a49a5b7c5c5faa604650e581/go/analysis/internal/checker/checker.go#L687.
We can do that in separate PR(s) though.